### ISSUE TITLE
Research LSTMs for processing larger amounts of data

### ISSUE THREAD

**Issue by MadeInGer** [sagr4019/ResearchProject#155]
We are currently struggling with the problem that our LSTM is not suitable for larger amounts of tokens. We want to research how other researcher are trying to solve this issue. This issue collects links to beneficial books/papers/articles that we have read.

**Comment by DanielEggelmann** [comment]
https://towardsdatascience.com/the-fall-of-rnn-lstm-2d1594c74ce0

**Comment by MadeInGer** [comment]
## http://colah.github.io/posts/2015-08-Understanding-LSTMs/

> LSTMs are explicitly designed to avoid the long-term dependency problem. Remembering information for long periods of time is practically their default behavior, not something they struggle to learn!

:arrow_right: I wonder why It’s not like that with us.





## Paper: LSTM: A Search Space Odyssey (https://arxiv.org/pdf/1503.04069.pdf)

The authors compared 8 variants of LSTMs with the result that there are no significant differences between those. However, they differ in performance: some are slightly faster.





## https://machinelearningmastery.com/handle-long-sequences-long-short-term-memory-recurrent-neural-networks/ and https://machinelearningmastery.com/truncated-backpropagation-through-time-in-keras/

Jason Brownlee lists 6 ways to handle long sequences with LSTMs.

> Modern recurrent neural networks like the Long Short-Term Memory, or LSTM, network are trained with a variation of the Backpropagation algorithm called Backpropagation Through Time. This algorithm has been modified further for efficiency on sequence prediction problems with very long sequences and is called Truncated Backpropagation Through Time. 





## https://danijar.com/tips-for-training-recurrent-neural-networks/

Danijar Hafner (PhD student) collected a few comprehensibly tips for training RNN, e. g. gradient clipping and truncated backpropagation for long sequences





# Points to ponder

- do we have the *vanishing gradient problem* (gradients multiplied by values less than 1 multiple times whereby they vanish) or the *exploding gradients* (gradients multiplied by values greater than 1 multiple times)? How do we know? It could be, since both are problems that appear at RNN and LSTM is a special kind of RNN. [Shamane Siriwardhana wrote something interesting](https://stackoverflow.com/questions/44478272/how-to-handle-extremely-long-lstm-sequence-length/44481566#44481566):

  - gradient clipping prevents this (upper bound to gradients)

  - truncated backpropagation through time

    > 1. Calculate the gradients only for number of given time steps 
    >    That means if your sequence is 200 time steps and you only give 10 time steps it will only calculate gradient for 10 time step and then pass the stored memory value in that 10 time step to next sequence(as the 
    >    initial cell state) .
    > 2. Take the full sequence and only back propagate gradients for some given time steps from selected time block. It's a continuous way 

- while reading [this post](https://datascience.stackexchange.com/a/31154) I asked my self: d0 we pad sequences to 0 (unused value) to ensure that all sequences have the same length?

**Comment by DanielEggelmann** [comment]
> * while reading [this post](https://datascience.stackexchange.com/a/31154) I asked my self: do we pad sequences to 0 (unused value) to ensure that all sequences have the same length?

Sequences are padded with 0, the value 0 is not used otherwise.

**Comment by DanielEggelmann** [comment]
> I came across this a couple of weeks ago, this gave me the idea, that the LSTM networks could be the problem.
> 
> https://towardsdatascience.com/the-fall-of-rnn-lstm-2d1594c74ce0

Reading through this again, the major problem with LSTM mentioned is performance.

> And one issue of RNN is that they are not hardware friendly. Let me explain: it takes a lot of resources we do not have to train these network fast. Also it takes much resources to run these model in the cloud, and given that the demand for speech-to-text is growing rapidly, the cloud is not scalable. We will need to process at the edge, right into the Amazon Echo! See note below for more details.

**Comment by DanielEggelmann** [comment]
[Gradient Flow in Recurrent Nets: the Difficulty of Learning Long-Term Dependencies](https://www.bioinf.jku.at/publications/older/ch7.pdf) This could be worth having a closer look, from what I understand, they are analysing the gradient behavior for recurrent networks.

**Comment by jan-christiansen** [comment]
[Fast Weight Long Short-Term Memory](https://arxiv.org/pdf/1804.06511) might be interesting because it discusses an abstract problem called ART, which is closely connected to remembering the types of all declared variables.
